i3lock.c

```diff
@@ -645,6 +645,120 @@ void handle_screen_resize(void) {
     redraw_screen();
 }
 
+static int read_raw_image_native(uint32_t *dest, FILE *src, int width, int height, int stride) {
+    int count = 0;
+    for (int y = 0; y < height; y++) {
+        size_t n = fread(&dest[y * stride], 1, width * 4, src);
+        count += n;
+        if (n < (size_t)(width * 4))
+            break;
+    }
+
+    return count;
+}
+
+static int read_raw_image_rgb(uint32_t *dest, FILE *src, int width, int height, int stride) {
+    unsigned char *buf = malloc(width * 3);
+    if (buf == NULL)
+        return -1;
+
+    int count = 0;
+    for (int y = 0; y < height; y++) {
+        size_t n = fread(buf, 1, width * 3, src);
+        count += n;
+        if (n < (size_t)(width * 3))
+            break;
+
+        for (int x = 0; x < width; ++x) {
+            int idx = x * 3;
+            dest[y * stride + x] = 0 |
+                                   (buf[idx + 0] << 16) |
+                                   (buf[idx + 1] << 8) |
+                                   (buf[idx + 2]);
+        }
+    }
+
+    free(buf);
+    return count;
+}
+
+static cairo_surface_t *read_raw_image(const char *image_path, const char *image_raw_format) {
+    cairo_surface_t *img;
+
+    /* Parse format as <width>x<height>[:<format>] */
+    int w = -1, h = -1;
+    char format[] = "native";
+    char *fmtstring = NULL;
+    if (asprintf(&fmtstring, "%%ix%%i:%%%lds", sizeof(format) - 1) == -1) {
+        errx(EXIT_FAILURE, "asprintf() failed.");
+    }
+
+    if (sscanf(image_raw_format, fmtstring, &w, &h, format) == EOF || w == -1 || h == -1) {
```

Comment:
What is so special about -1 that I cannot claim my image has width/height -1, but -2 is fine? :-P
(Yes, I know that this is meant to check for "did `sscanf` modify the argument at all?")